### PR TITLE
z/TPF Specific Build Updates for jilconst and ddr Executables

### PR DESCRIPTION
**Pull request by jjohnst-us**
Today zTPF build does not integrate jilconst and ddrgen
execution, and requires manual intervention to complete the build.
Other platforms are able to run these executables inline with
the rest of the build.  These updates will allow z/TPF to do the
same.

z/TPF is a cross compiled platform, that is to say development and
production environments are very differnt and what runs on the
development environment will not run on the production environment.
ddrgen and jilconst builds for other platforms assumethese systems
to be the same.  As part of this pull request we added the capability
for z/TPF development (openj9 builds) to launch a shell script which
will invoke a REST service to z/TPF which will load the appropriate
programs created in the middle of the build on the development
platform and run them.  Upon completion of the REST service z/tpf will
responsd with the generated artifacts and the development system can
continue with its build, as if it were done inline on the same
build platform.

Additionally this pull request has miscellaneous updates for both
cleanup and better integration with the z/TPF build environment.  For
example, we updated different path rules and variable labels to match
the rest of our operating system builds.

Signed-off-by: James D Johnston <jjohnst@us.ibm.com>

### CODE REVIEW

File: runtime/makelib/targets.mk.ftl [hunk 1]
```diff
@@ -242,8 +240,8 @@ endif
 UMA_LINK_PATH+=-L. -L$(UMA_PATH_TO_ROOT) -L$(UMA_PATH_TO_ROOT)lib/
 UMA_BUILD_LIBRARIES:=$(foreach lib,$(UMA_LIBRARIES),$(if $($(lib)_alllib),$(lib)))
 UMA_LINK_PATH+=$(foreach lib,$(UMA_BUILD_LIBRARIES),-L$($(lib)_path))
-# Remove duplicate paths
-UMA_LINK_PATH:=$(sort $(UMA_LINK_PATH))
+# remove sorting of UMA_LINK_PATH, changes link order
+#UMA_LINK_PATH:=$(sort $(UMA_LINK_PATH))
```
Comment (DanHeidinga) [inline runtime/makelib/targets.mk.ftl]: This changes linking on all unix platforms.  This change needs to be removed.

File: runtime/makelib/targets.mk.ftl [hunk 2]
```diff
@@ -531,7 +528,7 @@ cnathelp$(UMA_DOT_O):cnathelp.cpp
 	$(CXX) $(CXXFLAGS) -mstackrealign -c $<
 </#if>
 
-# JIT helper methods require us to disable buffer security checks on Windows
+# special handling for cnathelp.cpp to disable buffer security check in windows for JIT helpers
```
Comment (DanHeidinga) [inline runtime/makelib/targets.mk.ftl]: Please revert this change as it's not ztpf related.

File: runtime/makelib/targets.mk.ftl [hunk 3]
```diff
@@ -597,7 +594,7 @@ ifndef UMA_NO_CLEAN_TARGET
 #
 clean:
 	$(RM) $(UMA_OBJECTS) $(UMA_BYPRODUCTS) $(TARGETS)
-	-$(RM) *.d $(UMA_OBJECTS:$(UMA_DOT_O)=.i)
+	-$(RM) *.d $(UMA_OBJECTS:$(UMA_DOT_O)=.i) $(UMA_OBJECTS:$(UMA_DOT_O)=.lst)
```
Comment (DanHeidinga) [inline runtime/makelib/targets.mk.ftl]: Are `.lst` files ztpf specific?  If so the RM command should be only included on ztpf specs

File: runtime/vm/module.xml [hunk 4]
```diff
@@ -160,6 +160,9 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 			<library name="j9stackmap"/>
 			<library name="j9hookable"/>
 			<library name="j9omr" type="external"/>
+			<library name="omrglue" type="external">
```
Comment (DanHeidinga) [inline runtime/vm/module.xml]: Can you explain why this is necessary?  I don't understand why the vm would need to grow this dependency.

File: runtime/jilgen/module.xml [hunk 5]
```diff
@@ -59,6 +59,9 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 			<library name="j9pool" type="external"/>
 			<library name="j9exelib"/>
 			<library name="j9utilcore"/>
+			<library name="omrsig">
```
Comment (DanHeidinga) [inline runtime/jilgen/module.xml]: Can you explain why this is necessary?  It's surprising to see this linked to the signal chaining library

File: runtime/jilgen/module.xml [hunk 6]
```diff
@@ -98,6 +101,12 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 			<command line="cd $(UMA_PATH_TO_ROOT) &amp;&amp; qemu-arm -r '9' ./constgen" type="all">
 				<include-if condition="spec.linux_arm.*"/>
 			</command>
+			<command line="cd $(UMA_PATH_TO_ROOT)" type="all">
+				<include-if condition="spec.linux_ztpf.*"/>
+			</command>
+			<command line="if [ -x ../build_constgen.sh ] &amp;&amp; [ ! -e ../constgen.completed ] ; then ../build_constgen.sh ; fi" type="all">
```
Comment (DanHeidinga) [inline runtime/jilgen/module.xml]: Better to combine these two commands into a single one to ensure they run as group.

File: runtime/ddr/module.xml [hunk 7]
```diff
@@ -307,6 +307,9 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 			<library name="j9utilcore"/>
 			<library name="j9ddrautoblob"/>
 			<library name="j9gcautoblob"/>
+			<library name="omrsig">
```
Comment (DanHeidinga) [inline runtime/ddr/module.xml]: Can you explain why this is necessary? It's surprising to see this linked to the signal chaining library

File: runtime/ddr/module.xml [hunk 8]
```diff
@@ -335,18 +338,24 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
...
 				<exclude-if condition="spec.linux_arm.*"/>
+				<exclude-if condition="spec.linux_ztpf.*"/>
 			</command>
 			<command line="cd $(UMA_PATH_TO_ROOT) &amp;&amp; qemu-arm -r '9' ./j9ddrgen" type="all">
 				<include-if condition="spec.linux_arm.*"/>
 			</command>
+			<command line="cd $(UMA_PATH_TO_ROOT)" type="all">
+				<include-if condition="spec.linux_ztpf.*"/>
+			</command>
+			<command line="if [ -x ../build_j9ddrgen.sh ] &amp;&amp; [ ! -e ../j9ddrgen.completed ] ; then ../build_j9ddrgen.sh ; fi" type="all">
```
Comment (DanHeidinga) [inline runtime/ddr/module.xml]: Use a single command block for this.

File: runtime/makelib/targets.mk.ftl (hunk 2 above)
Comment (jjohnst-us) [reply runtime/makelib/targets.mk.ftl]: Reverted.

File: runtime/makelib/targets.mk.ftl (hunk 1 above)
Comment (jjohnst-us) [reply runtime/makelib/targets.mk.ftl]: Used spec file check to separate logic.

File: runtime/vm/module.xml (hunk 4 above)
Comment (jjohnst-us) [reply runtime/vm/module.xml]: Removed

File: runtime/jilgen/module.xml (hunk 5 above)
Comment (jjohnst-us) [reply runtime/jilgen/module.xml]: Removed.

File: runtime/jilgen/module.xml (hunk 6 above)
Comment (jjohnst-us) [reply runtime/jilgen/module.xml]: Done.

File: runtime/ddr/module.xml (hunk 7 above)
Comment (jjohnst-us) [reply runtime/ddr/module.xml]: Removed.

File: runtime/ddr/module.xml (hunk 8 above)
Comment (jjohnst-us) [reply runtime/ddr/module.xml]: Done.

**Comment by jjohnst-us** [comment]
@DanHeidinga 
Apologies, was creating PRs on behalf of someone else's updates.  Please let me know if you need anything else changed.

File: runtime/makelib/targets.mk.ftl (hunk 2 above)
Comment (pshipton) [reply runtime/makelib/targets.mk.ftl]: it doesn't look reverted

File: runtime/makelib/targets.mk.ftl [hunk 9]
```diff
@@ -513,7 +519,7 @@ MINGW_CXXFLAGS+=-std=c++0x -D_CRT_SUPPRESS_RESTRICT -DVS12AndHigher
 	MINGW_CXXFLAGS+=-D_M_X64
 </#if>
 endif
-# special handling BytecodeInterpreter.cpp and DebugBytecodeInterpreter.cpp
+# JIT helper methods require us to disable buffer security checks on Windows
```
Comment (pshipton) [inline runtime/makelib/targets.mk.ftl]: This comment change should be reverted as well.

Comment (jjohnst-us) [reply runtime/makelib/targets.mk.ftl]: @pshipton , misread 'changes' view.  Should be good now, also rebased.

File: runtime/makelib/targets.mk.ftl [hunk 10]
```diff
@@ -238,14 +236,23 @@ ifdef UMA_ADD_COPYRIGHT
...
 </#if>
 
+<#if uma.spec.type.ztpf>
+UMA_LINK_PATH+=-L. -L$(UMA_PATH_TO_ROOT) -L$(UMA_PATH_TO_ROOT)lib/
+UMA_BUILD_LIBRARIES:=$(foreach lib,$(UMA_LIBRARIES),$(if $($(lib)_alllib),$(lib)))
+UMA_LINK_PATH+=$(foreach lib,$(UMA_BUILD_LIBRARIES),-L$($(lib)_path))
+# removed sorting of UMA_LINK_PATH which is done for all unix platforms (see similar unix block above)
+# z/TPF unique build environment is sensitive to link order for some cases.
+#UMA_LINK_PATH:=$(sort $(UMA_LINK_PATH))
+</#if>
```
Comment (DanHeidinga) [inline runtime/makelib/targets.mk.ftl]: Rather than duplicating this whole block, can you #if just the sorting?

```
<#if uma.spec.type.unix>
UMA_LINK_PATH+=-L. -L$(UMA_PATH_TO_ROOT) -L$(UMA_PATH_TO_ROOT)lib/
UMA_BUILD_LIBRARIES:=$(foreach lib,$(UMA_LIBRARIES),$(if $($(lib)_alllib),$(lib)))
UMA_LINK_PATH+=$(foreach lib,$(UMA_BUILD_LIBRARIES),-L$($(lib)_path))
<#if !uma.spec.type.ztpf>
# Remove duplicate paths 
UMA_LINK_PATH:=$(sort $(UMA_LINK_PATH))
<#endif>
```

**Comment by DanHeidinga** [comment]
> Can you expand on how these changes help ztpf run jilconst and ddrgen inline?
>
> The way j9ddrgen.completed is used may be of interest to other platforms as well so documenting it here would be a plus.

@jjohnst-us are the ".completed" files something ztpf-specific?  It may be of benefit to other cross-compiled platforms.  Any explanation would be appreciated.

**Comment by jjohnst-us** [comment]
I updated the commit message with a description of the '.completed' files, it is used for preventing the REST request from being invoked multiple times in different phases of the build.  I collapsed the two uma blocks into one using the spec file as requested.

Found an issue with the z/TPF build where I merged a line incorrectly in combination with some in-lab tooling issues which took a long time to pinpoint.

**Comment by pshipton** [comment]
jenkins compile win jdk8

**Comment by pshipton** [comment]
jenkins test sanity plinux jdk8
